Fix #6400 QueuedThreadPool always interrupts threads in stop

gregw opened the pull request:
Fixes #6400 Even if there is no timeout, always interrupt pool threads to attempt to stop them.

gregw committed:
Fix #6400 QueuedThreadPool always interrupts threads in stop

Even if there is no timeout, always interrupt pool threads to attempt to stop them.

gregw commented:
I think it was an oversite that we don't at least interrupt threads.   We really wanted to stop threads in that situation, but that API is deprecated.

gregw committed:
Fix #6400 QueuedThreadPool always interrupts threads in stop

Fixed style

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -234,30 +234,29 @@ protected void doStop() throws Exception
-            joinThreads(System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout) / 2);
+        // Fill the job queue with noop jobs to wakeup idle threads.
+        for (int i = 0; i < threads; ++i)
+            jobs.offer(NOOP);
```
Since we are reviewing this, I don't think this loop guarantees that all threads are woken up.
On a single core, it may be that a single thread wakes up and consumes all the NOOP jobs in the queue.

gregw replied:
I think you are right.  The Runner used to have a `while(isRunning)` loop, so once woken it would check that state.  Now it is only checking the equivalent state only if it doesn't have a job.   So maybe we should always check the running state again?

gregw replied:
Ah no, the running state is already always tested - either in addCounts or explicitly if there is no jobs.  It might be clearer to always call addCounts.... let me see...

gregw replied:
@sbordet I've pushed a refactored version that I think makes this clearer... and I am 95% sure is exactly the same logic.
@lachlan-roberts can you have a look and triple check the latest checking doesn't actually change the idle thread count logic.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -234,30 +234,29 @@ protected void doStop() throws Exception
+            if (LOG.isDebugEnabled())
+                LOG.debug("Interrupting {}", thread);
+            thread.interrupt();
+        }
```
Now it may be possible that we interrupt threads immediately after submitting the NOOP jobs (or normal jobs) -- I'm not sure we want that if we want to gracefully stop (by letting the current tasks finish).

Also, interrupting is the right thing to do, but e.g. `ReservedThreadExecutor` does not handle it well: it just loops around and goes again into `take()/poll()`.

gregw replied:
Ah that is a problem. The `ReservedThread#run` method does do a `while(isRunning)`, but there is an inner loop in `ReservedThread#reservedWait` that doesn't.    I'll fix that....

gregw replied:
last check in fixed the inner loop of the ReservedThread

sbordet commented:
@gregw why does not QTP implements `Graceful`?

Mmmh. I know I have added review comments and I would like to improve, but I've been running in circles for a while now, and nothing seems better than what we have here 😃

gregw committed:
Fix #6400 QueuedThreadPool always interrupts threads in stop

 + Refactored exit condition for Runner main loop to make it clearer that it always checks running status
 + More comments to explain counts
 + Fixed inner loop of ReservedThread to ensure it always checks running status
 + Adjusted timing of the test to make it less flakey. Still a worry!

gregw committed:
Fix #6400 QueuedThreadPool always interrupts threads in stop

Added curlies
